packages/sdk/src/types/runtime/behaviors/clientAction.ts

```diff
@@ -0,0 +1,57 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Context } from '..';
+import { SetAudioStateOptions } from '../../..';
+import * as Payloads from '../../network/payloads';
+
+// tslint:disable: max-classes-per-file
+
+export abstract class ClientAction {
+	/** @hidden */
+	public abstract toJSON(): any;
+
+	/** @hidden */
+	public sendPayload(context: Context, targetId: string, actionName: string, actionState: string) {
+		context.internal.sendPayload({
+			type: 'set-client-action',
+			clientAction: {
+				...this.toJSON(),
+				targetId,
+				actionName,
+				actionState
+			}
+		} as Payloads.SetClientAction);
+	}
+}
+
+export class PlayAnimationAction extends ClientAction {
+	constructor(private animationName: string) {
+		super();
+	}
+
+	/** @hidden */
+	public toJSON() {
+		return {
+			type: 'play-animation',
```

Comment:
The type needs to be outside of the json here.  We need to know the type in the client before deserializing this blob, because we have to allocate the appropriate class to fill in.

Reply:
Untrue. In the deserializer, look at JObject['type'] before reading the rest of the values out of the JObject. This is how the CollisionGeometryConverter works.

---
In reply to: [316824678](https://github.com/microsoft/mixed-reality-extension-sdk/pull/425#discussion_r316824678) [](ancestors = 316824678)

Reply:
As I mentioned above.  Misread the converter code that I wrote a while back.  Disregard.